modules/stochastic_tools/python/stochastic/StochasticControl.py

```diff
@@ -0,0 +1,625 @@
+            physics_input (str): Filename of the physics simulation input.
+            parameters (list[str]): List of parameter names to sample.
+            quantities_of_interest (list[str]): List of QoI reporter names to collect.
+            options (StochasticRunOptions): Stochastic run configuration options.
+            physics_cli_args (list[str]): Additional CLI arguments for the physics app.
+            moose_port (int): Optional port number for web server control.
+            inherit_environment (bool): Whether or not the MOOSE command will
+                                        inherit the current shell environment.
+            poll_time (float): Polling time interval for checking MOOSE status.
+        """
+        self.physics_input: str = physics_input
+        self._parameters: list[str] = parameters
+        self.num_params: int = self._getNumParams(parameters)
+        self._qois: list[str] = quantities_of_interest
+        self.num_qois: int = len(self._qois)
+        self._opt: StochasticRunOptions = options
+        self._physics_cli_args: list[str] = physics_cli_args
+        self._input_file = os.path.abspath(self._opt.input_name)
+
+        # Build stochastic input
+        self._root: pyhit.Node = self._buildStochasticInput()
+
+        # Initialize base class
+        moose_command: list[str] = []
+        if self._opt.num_procs > 1:
+            moose_command += [self._opt.mpi_command, "-n", str(self._opt.num_procs)]
+        moose_command += [executable, "-i", self._opt.input_name]
+        moose_command += self._opt.cli_args
+        super().__init__(
+            moose_command=moose_command,
+            moose_control_name=self.web_server_control_name,
+            moose_port=moose_port,
+            inherit_environment=inherit_environment,
+            poll_time=poll_time,
+        )
+
+    def __enter__(self) -> "StochasticRunner":
+        """
+        Context manager entry to return a callable `StochasticRunner` object.
+        This enables easy usage in `with` blocks.
+
+        Unfortunately, we cannot initialize the simulation here because the run
+        needs a proper input matrix for the first time through.
+
+        Returns:
+            StochasticRunner: A callable runner instance.
+        """
+        return StochasticRunner(self)
+
+    def __exit__(self, exc_type=None, exc_value=None, ex_tb=None):
+        """
+        Clean up after a run by terminating or killing the MOOSE process
+        and removing the temporary input file.
+        """
+        # Terminate gracefully
+        if exc_type is None:
+            # If moose is waiting, send a terminate
+            try:
+                waiting_flag = self.getWaitingFlag()
+            except:
+                waiting_flag = None
+            if not waiting_flag is None:
+                self.setTerminate()
+
+            # Don't finalize if we never started
+            if self._initialized:
+                self.finalize()
```

Comment:
Can this logic not be somewhat recycled from the MooseControl?

Reply:
MooseControl will throw if it tries to finalize when the process is waiting. For this application, we want to terminate if the stochastic simulation is finished (the process will still be waiting) and then finalize.